### ISSUE TITLE
TypeErro Component.$__rerender

### ISSUE THREAD

**Issue by ovidius72** [marko-js/marko#647]
When I try to write something like the stop-watch component in the try online section, i get the following error when the script try to update the component:

```
Uncaught TypeError
    at Component.$__rerender (pages-1407d337.js:3344)
    at Component.update (pages-1407d337.js:3305)
    at updateComponents (pages-1407d337.js:1890)
    at updateUnbatchedComponents (pages-1407d337.js:1862)
    at pages-1407d337.js:1841
$__rerender @ pages-1407d337.js:3344
update @ pages-1407d337.js:3305
updateComponents @ pages-1407d337.js:1890
updateUnbatchedComponents @ pages-1407d337.js:1862
(anonymous) @ pages-1407d337.js:1841
```

below is the function where I hit the error: 

```
$__rerender: function(input) {
   if (input) {
      this.input = input;
   }
    var self = this;
    var renderer =  self.$__renderer;

    if (!renderer) {
        throw TypeError();
     }
.............................
```

`self.$__renderer` is undefined.

In my case there's a template rendered by express with marko/express.
The template has a tag called sidebar. This tag is where the code lives and it's build using the multi files structure:
component-browser.js
component.js
index.marko.

- component-browser.js handles the OnMount() and browser events.
- component.js handles onCreate()

In the onMount() I declared and assigned the intervalId to the global setTimeout() function that change the state using this.state.width every 1 second. (I'm using es6 function syntax so that the `this` scope is preserved)
When the interval changes the state i get the error i mentioned above.

Can you please help to understand where I'm wrong ?
Thanks

**Comment by maximilianschmitt** [comment]
I'm having the same issue on version 4.2.1.

**Comment by ovidius72** [comment]
@maximilianschmitt If you are using a main template you can try to something like this:

`template.marko`
```
include(input.lasso)
<!doctype html>
html
  head
    app-head(input)
    lasso-head
  body
    div id="main-app"
      app-header(input)
      app-body(input)

    browser-refresh enabled="true"
    lasso-body

  <!--custom scripts-->
  <include(input.scripts)/>
```

`page.marko`

```
include('~/theme/html.marko')
    @lasso
        lasso-page package-path="./browser.json"
    @title -- Dashboard
    @body
        app-main
```

Notice the @lasso tag in the page.marko file. Instead of putting the `lasso-page package-path=".."` tag in the main template I delegated the page to send it to the main template that receive it at the top of the file with the tage `include(input.lasso)`

I don't know why but that solved the issue for me.

**Comment by patrick-steele-idem** [comment]
**tldr; Don't use split UI components unless there is a really good reason to do so**

When using a split component (i.e., `component-browser.js` + `component.js`) then the JavaScript module bundler (e.g. Lasso.js) will only send down the `component-browser.js` and not `component.js` (assuming it is not nested within in an un-split component...). Split components allow the rendering logic to remain on the server (instead of also being sent to the browser). However, split UI components have the caveat that they cannot be rerendered unless the `component.js` is also sent down. You are seeing that error because `component.js` was not sent down to the browser and hence the rendering logic is missing in the browser for the UI component. The error message is cryptic and we should fix that (we need to improve error messages in development without adding bloat in production).

My advice would be to avoid using split UI components. Split UI components can have some benefits but they are usually a premature optimization with not too much gain.

I'm closing this issue because I think everything is working as intended (I'll open a separate issue for improving the error message). If you still have questions or concerns please feel free to add them here.

**Comment by maximilianschmitt** [comment]
Thanks for the clarification @patrick-steele-idem and @ovidius72! :)
